### DISCUSSION TITLE
2023-08-25 - Web-based Data Entry Forms UX - Chewing the Fat 📈

### DISCUSSION THREAD

**Discussion by piers-sinclair** [SSWConsulting/SSW.Rules.Content#6478] [Chewing the Fat]
Hey Form Designers!

Let's talk about [this rule](https://www.ssw.com.au/rules/data-entry-forms-for-web)

**Comment by jimmidier** [comment]
Great topic and rules! Another point I would add is for long and complex form it should be split into different pages based on data categories, making it more organized and easy to comprehend. 
![image](https://github.com/SSWConsulting/SSW.Rules.Content/assets/16027480/4cf4df41-3215-40dc-b707-546075a52e02)

**Reply by brydeno** [reply]
The contrast on this form is awful. Older users will suffer from severe eyestrain with the light grey on white and white on light grey. It also makes those things look disabled and I'm assuming they are not. This is a very simple accessibility issue to fix, but one that the industry as a whole seems to be moving actively away from fixing.

**Reply by Geordie88** [reply]
Agreed, I get the intention but the yellow section here is unpleasant

**Comment by tiagov8** [comment]
It's a great topic and Toby did very well on the video!

My feedback on the video content:

- 2:58 Disagree with placement of the fields to the right, generating a gap between them and labels
- ~~4:20 Would it be better to have the non-editable data in text only instead of a disabled field?~~
   **Update:** I talked to Adam and this makes sense as it is from a technical perspective, so I'm OK with it
- 5:35 It's weird how the good example for the responsiveness is a different screen from the bad example.. I don't feel it represents well what responsiveness should be 

- I think "Good" and "Bad" examples in the video should be in bold... and the word "example" in lowercase
- "Form States" heading makes more sense as "Clear button labels"
   **Update:** [The rule](https://www.ssw.com.au/rules/data-entry-forms-for-web/#use-descriptive-button-labels) has already been updated to reflect this
- "No Delete Button..." heading makes more sense as "Avoid "Delete" button..." 
   **Update:** [The rule](https://www.ssw.com.au/rules/data-entry-forms-for-web/#avoid-delete-button-for-item-lists) has already been updated to reflect this
- The title **"Mastering Web-Based Data Entry: A UX Handbook"** is possibly too pretentious... Maybe **"Tips for Better Web-Based Data Entry: A UX Exploration"** sounds more aligned with the content. UX is a huuuuuuge process that involves hundreds of aspects

---

IMPORTANT - I think every heading in the rule should be a separated rule. Currently it's talking about too many distinct points in the same rule... which is confusing for users and harder to maintain

From

> * [Do you know the correct way to develop web-based data entry forms?](https://www.ssw.com.au/rules/data-entry-forms-for-web/)

to

> * Do you have clear buttons and states in data entry forms?   
> 
> * Do you avoid "delete" butons for item lists in data entry forms?
> 
> * Do you validate data-entry forms and make them easy to fix?    
>    (should be an update on https://www.ssw.com.au/rules/do-you-provide-red-errors-next-to-the-field/)
> 
> * Do you care about data-entry fields formatting?
>    (should be related to https://ssw.com.au/rules/how-to-align-your-form-labels)
> 
> * Do you add "created" and "last updated by" information in data-entry forms?
> 
> * Do you set minimum defaults in data-entry fields?
> 
> * Do you make your data-entry forms responsive?    
>   (should be an update on https://ssw.com.au/rules/do-you-know-how-to-arrange-forms/)

---

[See my PR](https://github.com/SSWConsulting/SSW.Rules.Content/pull/6481) (approved by Adam) with improvements to the rule

**Reply by wicksipedia** [reply]
> 2:58 Disagree with placement to the right of the fields

@tiagov8 did you have an issue with the fields being
- text right aligned
- controls not full width
- spinner buttons on the right

Personally, I like the text being right aligned since it mimics what Excel does (and how we were taught to add numbers in school). Definitely makes it easier to spot order of magnitude mistakes.

The numeric controls did look odd not being a consistent width to the ones above though.

**Reply by tiagov8** [reply]
@wicksipedia 
I meant the gap between label and field... (I rephrased my comment to reflect that)

<img width="678" alt="Screenshot 2023-08-24 at 5 01 31 PM" src="https://github.com/SSWConsulting/SSW.Rules.Content/assets/12601228/e05301cc-eb9f-407d-aeb7-38abda356b91">

**Reply by tomek-i** [reply]
@tiagov8 I Agree with you but I also can't entirely agree :-P 
In this particular example, it is perfectly fine to fill out the gap and I would have done it too.

I'm sure you would agree that it depends on the overall width of the container and the general aesthetic and overall proportion, flow, etc.

Not in the above example but It would look weird having an input field that's 1024px wide to fill out available space that now would fit 1000 characters for a first name but realisticly it is never going to be more than 30 characters long. It would look off.

**Comment by AlexRothwell** [comment]
The rule is a great starting point for making user friendly forms. Of course accessibility is another aspect to be considered but while it's important, I can understand not including it here as it is a huge consideration with lots of nuance

**Comment by jakebayliss** [comment]
Great start to the rules on forms UX but agree with Tiago - probably needs to be split up as it seemed to jump between topics which could be their own more detailed rule

**Comment by 2fernandez** [comment]
I think it's an excellent rule and video for someone producing something like this for the first time.
Well done, Toby.

It needs a little bit of revision and brought back on track with UX principles.
But for the most part, the intention and the ideas of the topic are great

**Comment by tkapa** [comment]
Forms are simultaneously simple and complex. UX should be prioritised.

**Comment by ozdemir-mehmet** [comment]
Loved the mullet.

**Reply by wicksipedia** [reply]
That's our honey badger 👍

**Comment by MarcusIrmscherSSW** [comment]
It goes to show how important the UX is. Its not enough to just perform the task, it needs to be intuitive. (and fool proof) 😀

**Comment by vladislav-kir** [comment]
Its a very important topic, and I believe each section of the rule would work better as its own rule

**Comment by joshbermanssw** [comment]
Good UX makes data entry an easier process. Needs emphasis. (Also agree with the 'fool proof' aspect - auto-filled contents may need to be emphasised)

**Comment by Ben0189** [comment]
Great tips, especially for having create and last updated field for any editable data

**Comment by jeoffreyfischer** [comment]
Great summary of the good practice essentials for forms!

**Comment by Khaled-Albahsh** [comment]
Cool 👍

**Comment by cvionnet** [comment]
That's nice to have tips about UX!
The rule is a great source of information, and agree on the fact that it would be nice to split it up.

**Comment by Levijj22** [comment]
Great video, well done! @TobyChurches 
I agree with Tiago regarding the breakdown of rules.
The process of finding a specific rule should be more streamlined, allowing us to quickly locate the required item, rather than having to sift through a long rule.

**Comment by calumjs** [comment]
I look forward to the future where ChatGPT-powered forms allow for smart prefilling, validation, and error messages.
Maybe a text to speech/free text entry field on a page where you describe what you want to put in the form, then it fills out all the fields for you to double check.

**Comment by jackreimers** [comment]
Great rule, poor form design is a huge source of user frustration and can be the reason someone just stops using a product.
Need send this rule to the QLD government

**Comment by christoment** [comment]
Form is the worst dev experience but it is the most important part of an application!

Couple of feedback:
- Responsiveness - I'd always follow the Acceptance Criteria and/or the direction of the client - I think we should not add mobile friendly page unless it is agreed by the client, because responsiveness is a UX challenge as well - it's equal to make another desktop page.

- Show "created by" and "updated by" fields: I think this is a possible security leak - we should do this is a case-by-case scenario.

**Reply by tomek-i** [reply]
I agree that this should be explicitly confirmed with the client, might safe yourself a lot of work if its not required :-P 

I agree with point 2 as well, I think the example in the video is a little weak. 
I would argue in almost all cases the interacting user does not care about this information when filling out a form etc.
It perhaps becomes important in the "Admin" side of the project, where you set up defaults, configuration etc. That's where you potentially want to see who created it or set something up in a specific way.

**Comment by lukecookssw** [comment]
Great video Toby! You're a natural :)

**Comment by zacharykeeping** [comment]
I really like this video and rule - a lot of useful and essential tips to keep in mind when designing forms.

**Comment by bahjatalaadelssw** [comment]
I love auto populating forms! And don't make everything "required" stop stealing my data!!

**Comment by landonmaxwell** [comment]
Great video 👍

**Comment by chrisschultzssw** [comment]
This is important stuff. Great video too!

**Comment by amankumarrr** [comment]
I strongly believe that UX plays a pivotal role in enhancing the usability of your app. Regardless of how robust your backend is, without a strong UX, your app won't achieve the desired outcomes it truly deserves. We could benefit from more videos that delve into this topic.

**Comment by AntPolkanov** [comment]
Whenever a form is not working properly, especially larger ones, it totally gets on my nerves😡 I like the idea of scrolling to the field with failed validation or showing all the fields together at the top.

Thanks for the rule! 🙂

**Comment by jomelsontanco** [comment]
In FireBootCamp, we deal with a lot of form controls! Form Control UX is one of those silent background considerations where the measure of a "perfect" implementation is that no one notices, it just flows smoothly. On the other hand forgetting small details like not marking a required field or slightly misaligned input boxes can make a lot of difference in user experience. Very underrated topic!

**Comment by AshAnil** [comment]
Nice Video ⭐ UX plays a major role in a smooth user experience.

**Comment by Abf09** [comment]
Useful video considering I am working on form validations these days. It gave me some useful insights. 
One form that we can start implementing this rule on is the CTF for today which I was able to submit without anything for the 11. Important. Something as important would be useful to have on input like some of the other Questions there.

**Comment by JackDuan123** [comment]
I'm not a big fan of forms, I believe interaction can be simple and fine, everytime when I see a big form, my head start to ache.
A good UX can reduce the pain

**Comment by AttackOnMorty** [comment]
Great rule with detailed examples. A general guide for UX design is: **Don’t Make Me Think**

**Comment by Dhruv-0987** [comment]
Forms are the trickiest part of designing a UI and I believe the placing and the alignment of the input fields is important, because there is a lot of information on the screen at one time, the spacing and font size play a very important role in not letting the user get lost when they see a lot of text.
This rule is amazing and its a one stop for cross checking if your form is up to standards but need to look at the font and spacing as well in a form.

**Comment by jontygardner** [comment]
Great work @TobyChurches! Enjoyed learning about how to approach the UX of data entries. This applies particularly to the TV Team especially when clients fill out 'contact us' pages or 'feedback forms'. 

All the bad and good examples are awesome at visually showing this off. 

One noteworthy UX I've adopted in relation to submitting the date of the event - some people write their date dd/mm/yyyy or mm/dd/yyyy

This can complicate things if the user uses a different date method. 

Two solutions:

1. Make a calendar dropdown field

2. Make the date of the event a text field with an example placeholder for the user

E.g.

Date of event:
{{ Saturday, 7th of June 2025 }}

@bahjatalaadelssw love all the editing for this video! Great toucha adding in those camera movements to direct attention + the subtle sound effects for each tip. Let's use it for shots like the Newcastle building to setup the world :)

**Comment by steven0x51** [comment]
It's a great idea to introduce the topic of good UX design, but the rule has too many small fragmented points.

**Comment by Jord-Gui** [comment]
Great rule! Made me aware of things to look out for when designing forms.

**Comment by leahy268** [comment]
Good video and  rule.  Not something that I look at very often myself though

**Comment by JackDevAU** [comment]
Forms are needed on nearly every site. This rules gives some good recommendations to follow!

**Comment by leehawkinsssw** [comment]
The rule has some useful tips, but is too long and could be separated out into a number of smaller rules (@tiagov8 has done the hard yards in making good suggestions on this already).

Nice work by @TobyChurches on the video - the presentation is clear and well-paced.

**Comment by sylhuang** [comment]
Great rule and well-explained! Whenever I create a new form, I'm making arbitrary UX decisions based on my past experience. It is good to see tips documented which makes me re-think about many of my previous decisions and helps with consistency.

**Comment by tomek-i** [comment]
Is there a reason why the form does not indicate the required fields before the validation happens?
Usually, the required fields have a little `*` next to the label indicating that these are required, I've also seen forms with actual little annotations next to them like `(required)`.

I think it's a better experience to the user, especially on larger forms, to indicate what fields are required. 
![image](https://github.com/SSWConsulting/SSW.Rules.Content/assets/3699937/b58f3bbf-61a0-4f66-bf01-c94a4ab8be11)

**Comment by MattParkerDev** [comment]
Loved the video! UI/UX can often be overlooked.

**Comment by SamBWagner** [comment]
The video was radicool! 😎  

I think that forms are genuinely the reason most web devs quit the profession so having clear and concise helpful tips like this is really great!

**Comment by PennyWalker** [comment]
Great video Toby! Well done, thanks for presenting it in such a presentable way! :) UX is so important and it's great to see us doing some easy to action tips! :)

**Comment by babakamyljanovssw** [comment]
Great video that gave good tips to improve UX. Especially having 'Save' and 'Save and Close' buttons on form, so users can add multiple items in one way rather than going to form page many times or add one item and close the form. Sure will use this tip in my FireBootCamp project.

**Comment by drwharris** [comment]
Would like to see accessibility added to this.

**Reply by tomek-i** [reply]
Uff, that's another beast and probably deserves its own topic & series :-D

**Comment by stephanfakossw** [comment]
Another great rule! Great video Toby :-)
Not sure that I like the big gap in Figure: Good example - These fields have appropriate width for the data...Resizing of the boxes half way through does not feel natural which takes away my attention from content to formatting.

**Comment by BenDunkerleySSW** [comment]
Makes you realise how bad some forms are! Good video and good rule. Toby is very personable.

**Reply by leehawkinsssw** [reply]
An old favourite, but hasn't been updated for 20+ years http://hallofshame.gp.co.at/shame.htm Is there a modern version of this site as a source of even more bad examples?

**Reply by jomelsontanco** [reply]
this is a very funny one - https://goulartnogueira.github.io/BadUI/

**Reply by leehawkinsssw** [reply]
And don't do this (from a real website I used last week):

![image](https://github.com/SSWConsulting/SSW.Rules.Content/assets/106290015/c717a77f-2789-4cac-bc09-81a8565d8d0e)

**Reply by tomek-i** [reply]
I love the exceeding characters to drop into the bin :-D that's pretty cool

**Comment by jerryluossw** [comment]
Great rule, it is necessary to unify UI design specifications

**Comment by jaydenalchin** [comment]
Great stuff - and an important topic!

- **Buttons** - should also touch on button ordering, which button is the primary button and why (generally the primary action on the right indicates "forward progress" on a form)
    - Kind of mentioned here:  https://www.ssw.com.au/rules/do-you-make-your-cancel-button-less-obvious/
-  **Field formatting** - The 'bad' example feels more readable than the good one, due to some adjacent issues like text alignment, field size, label spacing and alignment
    - E.g. https://www.ssw.com.au/rules/how-to-align-your-form-labels/
    - Top-aligned labels are generally the best for readability and responsiveness despite being less space efficient at face value.
- **Validation** - live validations on blur or on input could be a cleaner solution, introducing higher user awareness since they see errors as soon as they're made.
- **Resizing** - love the concept and message but the examples here show a card, not a form? The second example used in the video shows an already small & narrow form that doesn't change notably at smaller breakpoints. Instead show left-aligned field labels wrapping into top-aligned ones at smaller screen size e.g. https://www.ssw.com.au/rules/do-you-know-how-to-arrange-forms/

All in all fantastic presentation and discussion. I agree with Tiago and others who've suggest this could be broken into smaller, specific, easier to find rules!

I would love to see accessibility touched on as well but that could/should be a whole other category of rules itself.

**Comment by jimmy1992abc** [comment]
Toby is such a good presenter/ speaker, explains the topic super clearly. I love the way to compare A and B on the same page, very straightforward.
It makes people crazy and super upset if someone spends 15 mins filling out the form and then being told that the form doesn't like it. A clear UI and design would definitely increase the UX and it would be the first step to gaining a positive view from the user.
Tip: The closing feels a little bit rushed, how about doing a retro so people know it is about to finish.

**Comment by Marxoz** [comment]
Good video and good information. Not all of us are UX / front end experts and needs some guidance on designs. 

I recently had a bad experience with uploading documents to a website. I had no idea that I had to upload all the documents, then choose what documents they are, and then press Submit. You only had one go. So every time I tried, I could only upload a single document before the link became invalid.

With a good UX design, it should have been clear that I need to upload all the documents before hitting submit, and I don't mean instructions. I mean the way that it looked visually. Will have to give them a call :)

**Comment by CoffeeBugHub** [comment]
The video was great! It talked about being careful while deleting records and mentioned adding the delete option when users edit a record. But I don't think that completely solves the issue. I think having a pop-up message to confirm the delete would work better.

**Comment by KristenHu** [comment]
Love the video, it's readable and useful! There are some UX rules I didn't notice before, it helps me a lot to have such guidance.

**Comment by ricksu978** [comment]
When comparing the Goods & Bad
I would keep it consistent, e.g: always side by side, bad on the left, good on the right or vice versa.

**Comment by Aibono1225** [comment]
Awesome video and great rules! They shared lots of good tips about UI/UX in a really clear way.

**Comment by AnkyChe** [comment]
Interesting topic and easy for a non-developer to understand. UI/UX is really important for users, I love the way the video showing bad examples along with good examples, easy to compare.

**Comment by hveraus** [comment]
Nice video Toby, a lot covered in a video under 7 minutes, well done. 
Some of my 2c below:
**Validation -** Actionable error messages should provided in context, in real-time, upon data entry, not after submit button clicked. 
**Button -** Cancel button should in a different style with the save button, and we can move it to the different side of the form.

**Comment by johnxu16** [comment]
It's a informative and interesting rule. Really looking forward to see more rules from designers. 
Really curious how many different ways we can remind users what kind of data they should input in the fields.
Ex. placeholder, icon

**Comment by tino-liu** [comment]
If there are a lot of information that needed to be gathered that means the amount of inputs increases, and for some people that might be annoying and daunting. So yeah, a good UI/UX design on forms definitely help solve some problems presented.

**Comment by sethdaily** [comment]
As a user, this is important. It's good to know the design principles that we follow for it.

**Comment by landonmaxwell** [comment]
![toby-ssw-newcastle](https://github.com/SSWConsulting/SSW.Rules.Content/assets/78135895/02667ce3-d447-42d8-89ca-31a0b62de8e0)

**Reply by landonmaxwell** [reply]
Petition for updated building art 😂

**Comment by BrookJeynes** [comment]
Nice rule and video. Form UI is an important aspect and can make or break whether someone bothers to fill it out or not. Small conveniences such as auto fill and instant validation goes a long way.

**Comment by tombui99** [comment]
This topic holds significant importance as for me, the majority of my prior client projects involved data entry forms. Therefore, knowing and being aware of these form UI/UX guidelines is crucial during developments.

**Comment by ozair-ashfaque** [comment]
Great tips in the video for better UX design.

**Comment by matt-goldman** [comment]
All enterprise applications are essentially collections of forms. Very important to get it right.

**Comment by danielmackay** [comment]
Great video and rule @TobyChurches.  Alot of common sense best practices all wrapped up into a single place.

I would be interested to know the design reasoning behind the numeric inputs not being as wide as the dropdowns?

**Comment by Harry-Ross** [comment]
Thought the video and rule were great introductions to UX design for forms. Would love to know more about UX for other elements (not forms) as well!

**Comment by Freego1783** [comment]
Important topic ! Nearly every app we make is using more or less complex forms, we have to know how to handle these.

**Comment by jernejk** [comment]
I'll go a bit old here and say, "In the old days, we done all of that and more. If users didn't like the UX, they uninstalled the app and installed another one. Only the best-engineered Windows Phone apps with the best UX stood a chance to survive the uninstall button! We used normal, bold, thin and semi-bold with up to 4 shades of black just to make the typography on the apps more like an art than a lazily written app!"

![OldManYellsAtSkySkyNewsGIF](https://github.com/SSWConsulting/SSW.Rules.Content/assets/5943653/e989727d-1c17-418f-8f87-30eb2eabc013)

Jokes aside, nice to see a bit more focus on UX is nice. :)

**Comment by Calvin-Lieu** [comment]
Very helpful rule and video to guide people with little frontend dev experience.
